README.md
MySQL Cookbook
=====================

The Mysql Cookbook is a library cookbook that provides resource primitives
(LWRPs) for use in recipes. It is designed to be a reference example for
creating highly reusable cross-platform cookbooks.

Scope
-----
This cookbook is concerned with the "MySQL Community Server",
particularly those shipped with F/OSS Unix and Linux distributions. It
does not address forks or value-added repackaged MySQL distributions
like Drizzle, MariaDB, or Percona.

Requirements
------------
- Chef 11 or higher
- Ruby 1.9 or higher (preferably from the Chef full-stack installer)
- Network accessible package repositories
- 'recipe[selinux::disabled]' on RHEL platforms

Platform Support
----------------
The following platforms have been tested with Test Kitchen:

```
|----------------+-----+-----+-----+-----+-----|
|                | 5.0 | 5.1 | 5.5 | 5.6 | 5.7 |
|----------------+-----+-----+-----+-----+-----|
| debian-7       |     |     | X   |     |     |
|----------------+-----+-----+-----+-----+-----|
| ubuntu-10.04   |     | X   |     |     |     |
|----------------+-----+-----+-----+-----+-----|
| ubuntu-12.04   |     |     | X   |     |     |
|----------------+-----+-----+-----+-----+-----|
| ubuntu-14.04   |     |     | X   | X   |     |
|----------------+-----+-----+-----+-----+-----|
| centos-5       |   X | X   | X   | X   | X   |
|----------------+-----+-----+-----+-----+-----|
| centos-6       |     | X   | X   | X   | X   |
|----------------+-----+-----+-----+-----+-----|
| centos-7       |     |     | X   | X   | X   |
|----------------+-----+-----+-----+-----+-----|
| amazon         |     |     | X   | X   | X   |
|----------------+-----+-----+-----+-----+-----|
| fedora-20      |     |     | X   | X   | X   |
|----------------+-----+-----+-----+-----+-----|
| suse-11.3      |     |     | X   |     |     |
|----------------+-----+-----+-----+-----+-----|
| omnios-151006  |     |     | X   | X   |     |
|----------------+-----+-----+-----+-----+-----|
| smartos-14.3.0 |     |     | X   | X   |     |
|----------------+-----+-----+-----+-----+-----|
```

Cookbook Dependencies
------------
- yum-mysql-community
- smf

Usage
-----
Place a dependency on the mysql cookbook in your cookbook's metadata.rb
```ruby
depends 'mysql', '~> 6.0'
```

Then, in a recipe:

```ruby
mysql_service 'default' do
  port '3306'
  version '5.5'
  initial_root_password 'change me'
  action [:create, :start]
end

mysql_config 'default' do
  source 'mysite.cnf.erb'
  notifies :restart, 'mysql_service[default]'
  action :create
end
```

Resources Overview
------------------
### mysql_service

The `mysql_service` resource manages the basic plumbing needed to get a
MySQL server instance running with minimal configuration.

The `:create` action handles package installation, support
directories, socket files, and other operating system level concerns.
The internal configuration file contains just enough to get the
service up and running, then loads extra configuration from a conf.d
directory. Further configurations are managed with the `mysql_config` resource.

- If the `data_dir` is empty, a database will be initialized, and a
root user will be set up with `initial_root_password`. If this
directory already contains database files, no action will be taken.

The `:start` action starts the service on the machine using the
appropriate provider for the platform. The `:start` action should be
omitted when used in recipes designed to build containers.

#### Example
```ruby
mysql_service 'default' do
  version '5.7'
  bind_address '0.0.0.0'
  port '3306'  
  data_dir '/data'
  initial_root_password 'Ch4ng3me'
  action [:create, :start]
end
```

Please note that when using `notifies` or `subscribes`, the resource
to reference is `mysql_service[name]`, not `service[mysql]`.

#### Parameters

- `charset` - specifies the default character set. Defaults to `utf8`.

- `data_dir` - determines where the actual data files are kept
on the machine. This is useful when mounting external storage. When
omitted, it will default to the platform's native location.

- `initial_root_password` - allows the user to specify the initial
  root password for mysql when initializing new databases.
  This can be set explicitly in a recipe, driven from a node
  attribute, or from data_bags. When omitted, it defaults to
  `ilikerandompasswords`. Please be sure to change it.

- `instance` - A string to identify the MySQL service. By convention,
  to allow for multiple instances of the `mysql_service`, directories
  and files on disk are named `mysql-<instance_name>`. Defaults to the
  resource name.

- `package_action` - Defaults to `:install`.

- `package_name` - Defaults to a value looked up in an internal map.

- `package_version` - Specific version of the package to install,
  passed onto the underlying package manager. Defaults to `nil`.

- `bind_address` - determines the listen IP address for the mysqld service. When
  omitted, it will be determined by MySQL. If the address is "regular" IPv4/IPv6
  address (e.g 127.0.0.1 or ::1), the server accepts TCP/IP connections only for
  that particular address. If the address is "0.0.0.0" (IPv4) or "::" (IPv6), the
  server accepts TCP/IP connections on all IPv4 or IPv6 interfaces.

- `port` - determines the listen port for the mysqld service. When
  omitted, it will default to '3306'.

- `run_group` - The name of the system group the `mysql_service`
  should run as. Defaults to 'mysql'.

- `run_user` - The name of the system user the `mysql_service` should
  run as. Defaults to 'mysql'.

- `socket` - determines where to write the socket file for the
  `mysql_service` instance. Useful when configuring clients on the
  same machine to talk over socket and skip the networking stack.
  Defaults to a calculated value based on platform and instance name.

- `version` - allows the user to select from the versions available
  for the platform, where applicable. When omitted, it will install
  the default MySQL version for the target platform. Available version
  numbers are `5.0`, `5.1`, `5.5`, `5.6`, and `5.7`, depending on platform.

#### Actions

- `:create` - Configures everything but the underlying operating system service.
- `:delete` - Removes everything but the package and data_dir.
- `:start` - Starts the underlying operating system service
- `:stop`-  Stops the underlying operating system service
- `:restart` - Restarts the underlying operating system service
- `:reload` - Reloads the underlying operating system service

#### Providers
Chef selects the appropriate provider based on platform and version,
but you can specify one if your platform support it.

```ruby
mysql_service[instance-1] do
  port '1234'
  data_dir '/mnt/lottadisk'
  provider Chef::Provider::MysqlService::Sysvinit
  action [:create, :start]
end
```

- `Chef::Provider::MysqlService` - Configures everything needed t run
a MySQL service except the platform service facility. This provider
should never be used directly. The `:start`, `:stop`, `:restart`, and
`:reload` actions are stubs meant to be overridden by the providers
below.

- `Chef::Provider::MysqlService::Smf` - Starts a `mysql_service` using
the Service Management Facility, used by Solaris and IllumOS. Manages
the FMRI and method script.

- `Chef::Provider::MysqlService::Systemd` - Starts a `mysql_service`
using SystemD. Manages the unit file and activation state

- `Chef::Provider::MysqlService::Sysvinit` - Starts a `mysql_service`
using SysVinit. Manages the init script and status.

- `Chef::Provider::MysqlService::Upstart` - Starts a `mysql_service`
using Upstart. Manages job definitions and status.

### mysql_config

The `mysql_config` resource is a wrapper around the core Chef
`template` resource. Instead of a `path` parameter, it uses the
`instance` parameter to calculate the path on the filesystem where
file is rendered.

#### Example

```ruby
mysql_config[default] do
  source 'site.cnf.erb'
  action :create
end
```

#### Parameters

- `config_name` - The base name of the configuration file to be
  rendered into the conf.d directory on disk. Defaults to the resource
  name.

- `cookbook` - The name of the cookbook to look for the template
  source. Defaults to nil

- `group` - System group for file ownership. Defaults to 'mysql'.

- `instance` - Name of the `mysql_service` instance the config is
  meant for. Defaults to 'default'.

- `owner` - System user for file ownership. Defaults to 'mysql'.

- `source` - Template in cookbook to be rendered.

- `variables` - Variables to be passed to the underlying `template`
  resource.

- `version` - Version of the `mysql_service` instance the config is
  meant for. Used to calculate path. Only necessary when using
  packages with unique configuration paths, such as RHEL Software
  Collections or OmniOS. Defaults to 'nil'

#### Actions
- `:create` - Renders the template to disk at a path calculated using
  the instance parameter.
  
- `:delete` - Deletes the file from the conf.d directory calculated
  using the instance parameter.

#### More Examples
```ruby
mysql_service 'instance-1' do
  action [:create, :start]
end

mysql_service 'instance-2' do
  action [:create, :start]
end

mysql_config 'logging' do
  instance 'instance-1'
  source 'logging.cnf.erb'
  action :create
  notifies :restart, 'mysql_service[instance-1]'
end

mysql_config 'security settings for instance-2' do
  config_name 'security'
  instance 'instance-2'
  source 'security_stuff.cnf.erb'
  variables(:foo => 'bar')
  action :create
  notifies :restart, 'mysql_service[instance-2]'
end
```

### mysql_client
The `mysql_client` resource manages the MySQL client binaries and
development libraries. 

It is an example of a "singleton" resource. Declaring two
`mysql_client` resources on a machine usually won't yield two separate
copies of the client binaries, except for platforms that support
multiple versions (RHEL SCL, OmniOS).

#### Example
```ruby
mysql_client 'default' do
  action :create
end
```

#### Parameters
- `package_name` - An array of packages to be installed. Defaults to a
  value looked up in an internal map.  

- `package_version` - Specific versions of the package to install,

CHANGELOG.md
mysql Cookbook CHANGELOG
========================

v6.0.20 (2015-03-27)
--------------------
- #318 - Fixing Upstart pre-start script to handle custom socket paths

v6.0.19 (2015-03-25)
--------------------
- Adding support for Amazon Linux 2015.03

v6.0.18 (2015-03-24)
--------------------
- Adding support for 5.6 and 5.7 packages from dotdeb repos on Debian 7

v6.0.17 (2015-03-13)
--------------------
- Updated for MySQL 5.7.6.
- Handing removal of mysql_install_db and mysqld_safe

v6.0.16 (2015-03-10)
--------------------
- Moved --defaults-file as first option to mysql_install_db_script

v6.0.15 (2015-02-26)
--------------------
- Updating docker detection fix to pass specs

v6.0.14 (2015-02-26)
--------------------
- Fixed debian system service :disable action. Now survives reboot
- Fixing centos-7 instance :enable action. Now survives
- Not applying Apparmor policy if running in a Docker container

v6.0.13 (2015-02-15)
--------------------
- Adding support for special characters in initial_root_password
- Fixing failure status bug in sysvinit script

v6.0.12 (2015-02-30)
--------------------
- No changes. Released a 6.0.11 that was identical to 6.0.10.
  Git before coffee.

v6.0.11 (2015-02-30)
--------------------
- Adding support for configurable socket files

v6.0.10 (2015-01-19)
------------------
- Fix #282 - Fixing up data_dir template variable

v6.0.9 (2015-01-19)
------------------
- Fix #282 - undefined method `parsed_data_dir' bug

v6.0.8 (2015-01-19)
------------------
- Refactoring helper methods out of resource classes

v6.0.7 (2015-01-14)
------------------
- Fixing timing issue with Upstart provider :restart and :reload
  actions where service returns before being available

v6.0.6 (2014-12-26)
------------------
- Fixing subtle bug where MysqlCookbook::Helper methods were polluting Chef::Resource

v6.0.5 (2014-12-25)
------------------
- Using 'include_recipe' instead of 'recipe_eval' in LWRP
- Fixing type checking on package_name attribute in mysql_client resource.

v6.0.4 (2014-12-21)
------------------
- Suggest available versions if current is not available for current platform.

v6.0.3 (2014-12-17)
------------------
- Adding bind_address parameter to mysql_service resource

v6.0.2 (2014-12-17)
------------------
- Fixing sysvinit provider to survive reboots

v6.0.1 (2014-12-16)
------------------
- Fixing Upstart template to survive reboots

v6.0.0 (2014-12-15)
------------------
- Major version update
- Cookbook now provides LWRPs instead of recipes
- Platform providers re-factored into init system providers
- Separated :create and :start actions for use in recipes that build containers
- mysql_service now supports multiple instances on the same machine
- mysql_service no longer attempts to manage user records
- Removal of debian-sys-maint
- Unified Sysvinit script that works on all platforms
- mysql_config resource introduced
- mysql_client fixed up
- Refactored acceptance tests
- Temporarily dropped FreeBSD support

v5.6.1 (2014-10-29)
------------------
- Use Gem::Version instead of Chef::Version

v5.6.0 (2014-10-29)
------------------
- Changing default charset to utf8
- Quoting passwords in debian.cnf.erb
- Amazon 2014.09 support
- Ubuntu 14.10 support
- Only hide passwords from STDOUT via "sensitive true" in chef-client higher than 11.14
- Updating test harness

v5.5.4 (2014-10-07)
------------------
- Adding sensitive flag to execute resources to protect passwords from logs  

v5.5.3 (2014-09-24)
------------------
- Reverting back to Upstart on Ubuntu 14.04

v5.5.2 (2014-09-8)
------------------
- Reverting commit that broke Debian pass_string

v5.5.1 (2014-09-2)
------------------
- Switching Ubuntu service provider to use SysVinit instead of Upstart

v5.5.0 (2014-08-27)
-------------------
- Adding package version and action parameters to mysql_service resource
- Fixing Debian pass_string

v5.4.4 (2014-08-27)
-------------------
- Changing module namespace to MysqlCookbook

v5.4.3 (2014-08-25)
-------------------
- More refactoring. Moving helper function bits into resource parsed_parameters

v5.4.2 (2014-08-25)
-------------------
- Moving provider local variables into definitions for RHEL provider

v5.4.1 (2014-08-25)
-------------------
- Refactoring resources into the LWRP style with parsed parameters
- Moving provider local variables into definitions

v5.4.0 (2014-08-25)
-------------------
- #212 - support for centos-7 (mysql55 and mysql56)
- Adding (untested) Debian-6 support
- Adding Suse support to metadata.rb
- Adding ability to change MySQL root password
- Added libmysqlclient-devel package to SuSE client provider
- Appeasing AppArmor
- Reducing duplication in client provider

v5.3.6 (2014-06-18)
-------------------
- Fixing pid path location. Updating tests to include real RHEL


v5.3.4 (2014-06-16)
-------------------
- Fixing specs for Amazon Linux server package names


v5.3.2 (2014-06-16)
-------------------
- Fixing Amazon Linux support


v5.3.0 (2014-06-11)
-------------------
- #189 - Fix server_repl_password description

libraries/helpers.rb
require 'shellwords'

module MysqlCookbook
  module Helpers
    include Chef::DSL::IncludeRecipe

    def base_dir
      prefix_dir || '/usr'
    end

    def configure_package_repositories
      # we need to enable the yum-mysql-community repository to get packages
      return unless %w(rhel fedora).include? node['platform_family']
      case parsed_version
      when '5.5'
        # Prefer packages from native repos
        return if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 5
        return if node['platform_family'] == 'fedora'
        include_recipe('yum-mysql-community::mysql55')
      when '5.6'
        include_recipe('yum-mysql-community::mysql56')
      when '5.7'
        include_recipe('yum-mysql-community::mysql57')
      end
    end

    def client_package_name
      return new_resource.package_name if new_resource.package_name
      client_package
    end

    def defaults_file
      "#{etc_dir}/my.cnf"
    end

    def error_log
      "#{log_dir}/error.log"
    end

    def etc_dir
      return "/opt/mysql#{pkg_ver_string}/etc/#{mysql_name}" if node['platform_family'] == 'omnios'
      return "#{prefix_dir}/etc/#{mysql_name}" if node['platform_family'] == 'smartos'
      "#{prefix_dir}/etc/#{mysql_name}"
    end

    def include_dir
      "#{etc_dir}/conf.d"
    end

    def lc_messages_dir
    end

    def log_dir
      return "/var/adm/log/#{mysql_name}" if node['platform_family'] == 'omnios'
      "#{prefix_dir}/var/log/#{mysql_name}"
    end

    def mysql_name
      "mysql-#{new_resource.instance}"
    end

    def pkg_ver_string
      parsed_version.gsub('.', '') if node['platform_family'] == 'omnios'
    end

    def prefix_dir
      return "/opt/mysql#{pkg_ver_string}" if node['platform_family'] == 'omnios'
      return '/opt/local' if node['platform_family'] == 'smartos'
      return "/opt/rh/#{scl_name}/root" if scl_package?
    end

    def scl_name
      return unless node['platform_family'] == 'rhel'
      return 'mysql51' if parsed_version == '5.1' && node['platform_version'].to_i == 5
      return 'mysql55' if parsed_version == '5.5' && node['platform_version'].to_i == 5
    end

    def scl_package?
      return unless node['platform_family'] == 'rhel'
      return true if parsed_version == '5.1' && node['platform_version'].to_i == 5
      return true if parsed_version == '5.5' && node['platform_version'].to_i == 5
      false
    end

    def system_service_name
      return 'mysql51-mysqld' if node['platform_family'] == 'rhel' && scl_name == 'mysql51'
      return 'mysql55-mysqld' if node['platform_family'] == 'rhel' && scl_name == 'mysql55'
      return 'mysqld' if node['platform_family'] == 'rhel'
      return 'mysqld' if node['platform_family'] == 'fedora'
      return 'mysql' if node['platform_family'] == 'debian'
      return 'mysql' if node['platform_family'] == 'suse'
      return 'mysql' if node['platform_family'] == 'omnios'
      return 'mysql' if node['platform_family'] == 'smartos'
    end

    def v56plus
      return false if parsed_version.split('.')[0].to_i < 5
      return false if parsed_version.split('.')[1].to_i < 6
      true
    end

    def v57plus
      return false if parsed_version.split('.')[0].to_i < 5
      return false if parsed_version.split('.')[1].to_i < 7
      true
    end

    # database and initial records
    # initialization commands

    def mysqld_initialize_cmd
      cmd = mysqld_bin
      cmd << " --defaults-file=#{etc_dir}/my.cnf"
      cmd << ' --initialize'
      cmd << ' --explicit_defaults_for_timestamp' if v56plus
      return "scl enable #{scl_name} \"#{cmd}\"" if scl_package?
      cmd
    end

    def mysql_install_db_cmd
      cmd = mysql_install_db_bin
      cmd << " --defaults-file=#{etc_dir}/my.cnf"
      cmd << " --datadir=#{parsed_data_dir}"
      return "scl enable #{scl_name} \"#{cmd}\"" if scl_package?
      cmd
    end

    def record_init
      cmd = v56plus ? mysqld_bin : mysqld_safe_bin
      cmd << " --defaults-file=#{etc_dir}/my.cnf"
      cmd << " --init-file=/tmp/#{mysql_name}/my.sql"
      cmd << ' --explicit_defaults_for_timestamp' if v56plus
      cmd << ' &'
      return "scl enable #{scl_name} \"#{cmd}\"" if scl_package?
      cmd
    end

    def db_init
      return mysqld_initialize_cmd if v57plus
      mysql_install_db_cmd
    end

    def init_records_script
      <<-EOS
        set -e
        rm -rf /tmp/#{mysql_name}
        mkdir /tmp/#{mysql_name}

        cat > /tmp/#{mysql_name}/my.sql <<-EOSQL
DELETE FROM mysql.user ;
CREATE USER 'root'@'%' IDENTIFIED BY '#{Shellwords.escape(new_resource.initial_root_password)}' ;
GRANT ALL ON *.* TO 'root'@'%' WITH GRANT OPTION ;
FLUSH PRIVILEGES;
DROP DATABASE IF EXISTS test ;
EOSQL

       #{db_init}
       #{record_init}

       while [ ! -f #{pid_file} ] ; do sleep 1 ; done
       kill `cat #{pid_file}`
       while [ -f #{pid_file} ] ; do sleep 1 ; done
       rm -rf /tmp/#{mysql_name}
       EOS
    end

    def mysql_bin
      return "#{prefix_dir}/bin/mysql" if node['platform_family'] == 'smartos'
      return "#{base_dir}/bin/mysql" if node['platform_family'] == 'omnios'
      "#{prefix_dir}/usr/bin/mysql"
    end

    def mysql_install_db_bin
      return "#{base_dir}/scripts/mysql_install_db" if node['platform_family'] == 'omnios'
      return "#{prefix_dir}/bin/mysql_install_db" if node['platform_family'] == 'smartos'
      'mysql_install_db'
    end

    def mysql_version
      new_resource.version
    end

    def mysqladmin_bin
      return "#{prefix_dir}/bin/mysqladmin" if node['platform_family'] == 'smartos'
      return 'mysqladmin' if scl_package?
      "#{prefix_dir}/usr/bin/mysqladmin"
    end

    def mysqld_bin
      return "#{prefix_dir}/libexec/mysqld" if node['platform_family'] == 'smartos'
      return "#{base_dir}/bin/mysqld" if node['platform_family'] == 'omnios'
      return '/usr/sbin/mysqld' if node['platform_family'] == 'fedora' && v56plus
      return '/usr/libexec/mysqld' if node['platform_family'] == 'fedora'
      return 'mysqld' if scl_package?
      "#{prefix_dir}/usr/sbin/mysqld"
    end

    def mysqld_safe_bin
      return "#{prefix_dir}/bin/mysqld_safe" if node['platform_family'] == 'smartos'
      return "#{base_dir}/bin/mysqld_safe" if node['platform_family'] == 'omnios'
      return 'mysqld_safe' if scl_package?
      "#{prefix_dir}/usr/bin/mysqld_safe"
    end

    def pid_file
      "#{run_dir}/mysqld.pid"
    end

    def run_dir
      return "#{prefix_dir}/var/run/#{mysql_name}" if node['platform_family'] == 'rhel'
      return "/run/#{mysql_name}" if node['platform_family'] == 'debian'
      "/var/run/#{mysql_name}"
    end

    def sensitive_supported?
      Gem::Version.new(Chef::VERSION) >= Gem::Version.new('11.14.0')
    end

    def socket_file
      return new_resource.socket if new_resource.socket
      "#{run_dir}/mysqld.sock"
    end

    def socket_dir
      return File.dirname(new_resource.socket) if new_resource.socket
      run_dir
    end

    def tmp_dir
      '/tmp'
    end

    #######
    # FIXME: There is a LOT of duplication here..
    # There has to be a less gnarly way to look up this information. Refactor for great good!
    #######
    class Pkginfo
      def self.pkginfo
        # Autovivification is Perl.
        @pkginfo = Chef::Node.new

        @pkginfo.set['debian']['10.04']['5.1']['client_package'] = %w(mysql-client-5.1 libmysqlclient-dev)
        @pkginfo.set['debian']['10.04']['5.1']['server_package'] = 'mysql-server-5.1'
        @pkginfo.set['debian']['12.04']['5.5']['client_package'] = %w(mysql-client-5.5 libmysqlclient-dev)
        @pkginfo.set['debian']['12.04']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['13.04']['5.5']['client_package'] = %w(mysql-client-5.5 libmysqlclient-dev)
        @pkginfo.set['debian']['13.04']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['13.10']['5.5']['client_package'] = %w(mysql-client-5.5 libmysqlclient-dev)
        @pkginfo.set['debian']['13.10']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['14.04']['5.5']['client_package'] = %w(mysql-client-5.5 libmysqlclient-dev)
        @pkginfo.set['debian']['14.04']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['14.04']['5.6']['client_package'] = %w(mysql-client-5.6 libmysqlclient-dev)
        @pkginfo.set['debian']['14.04']['5.6']['server_package'] = 'mysql-server-5.6'
        @pkginfo.set['debian']['14.10']['5.5']['client_package'] = %w(mysql-client-5.5 libmysqlclient-dev)
        @pkginfo.set['debian']['14.10']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['14.10']['5.6']['client_package'] = %w(mysql-client-5.6 libmysqlclient-dev)
        @pkginfo.set['debian']['14.10']['5.6']['server_package'] = 'mysql-server-5.6'
        @pkginfo.set['debian']['6']['5.1']['client_package'] = %w(mysql-client libmysqlclient-dev)
        @pkginfo.set['debian']['6']['5.1']['server_package'] = 'mysql-server-5.1'
        @pkginfo.set['debian']['7']['5.5']['client_package'] = %w(mysql-client libmysqlclient-dev)
        @pkginfo.set['debian']['7']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['debian']['7']['5.6']['client_package'] = %w(mysql-client libmysqlclient-dev) # apt-repo from dotdeb
        @pkginfo.set['debian']['7']['5.6']['server_package'] = 'mysql-server-5.6'
        @pkginfo.set['debian']['7']['5.7']['client_package'] = %w(mysql-client libmysqlclient-dev) # apt-repo from dotdeb
        @pkginfo.set['debian']['7']['5.7']['server_package'] = 'mysql-server-5.7'
        @pkginfo.set['debian']['jessie/sid']['5.5']['client_package'] = %w(mysql-client libmysqlclient-dev)
        @pkginfo.set['debian']['jessie/sid']['5.5']['server_package'] = 'mysql-server-5.5'
        @pkginfo.set['fedora']['20']['5.5']['client_package'] = %w(community-mysql community-mysql-devel)
        @pkginfo.set['fedora']['20']['5.5']['server_package'] = 'community-mysql-server'
        @pkginfo.set['fedora']['20']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['fedora']['20']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['fedora']['20']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['fedora']['20']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['freebsd']['10']['5.5']['client_package'] = %w(mysql55-client)
        @pkginfo.set['freebsd']['10']['5.5']['server_package'] = 'mysql55-server'
        @pkginfo.set['freebsd']['9']['5.5']['client_package'] = %w(mysql55-client)
        @pkginfo.set['freebsd']['9']['5.5']['server_package'] = 'mysql55-server'
        @pkginfo.set['omnios']['151006']['5.5']['client_package'] = %w(database/mysql-55/library)
        @pkginfo.set['omnios']['151006']['5.5']['server_package'] = 'database/mysql-55'
        @pkginfo.set['omnios']['151006']['5.6']['client_package'] = %w(database/mysql-56)
        @pkginfo.set['omnios']['151006']['5.6']['server_package'] = 'database/mysql-56'
        @pkginfo.set['rhel']['2014.09']['5.1']['server_package'] = %w(mysql51 mysql51-devel)
        @pkginfo.set['rhel']['2014.09']['5.1']['server_package'] = 'mysql51-server'
        @pkginfo.set['rhel']['2014.09']['5.5']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2014.09']['5.5']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['2014.09']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2014.09']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['2014.09']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2014.09']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['2015.03']['5.1']['server_package'] = %w(mysql51 mysql51-devel)
        @pkginfo.set['rhel']['2015.03']['5.1']['server_package'] = 'mysql51-server'
        @pkginfo.set['rhel']['2015.03']['5.5']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2015.03']['5.5']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['2015.03']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2015.03']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['2015.03']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['2015.03']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['5']['5.0']['client_package'] = %w(mysql mysql-devel)
        @pkginfo.set['rhel']['5']['5.0']['server_package'] = 'mysql-server'
        @pkginfo.set['rhel']['5']['5.1']['client_package'] = %w(mysql51-mysql)
        @pkginfo.set['rhel']['5']['5.1']['server_package'] = 'mysql51-mysql-server'
        @pkginfo.set['rhel']['5']['5.5']['client_package'] = %w(mysql55-mysql mysql55-mysql-devel)
        @pkginfo.set['rhel']['5']['5.5']['server_package'] = 'mysql55-mysql-server'
        @pkginfo.set['rhel']['5']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['5']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['5']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['5']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['6']['5.1']['client_package'] = %w(mysql mysql-devel)
        @pkginfo.set['rhel']['6']['5.1']['server_package'] = 'mysql-server'
        @pkginfo.set['rhel']['6']['5.5']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['6']['5.5']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['6']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['6']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['6']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['6']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['7']['5.5']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['7']['5.5']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['7']['5.6']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['7']['5.6']['server_package'] = 'mysql-community-server'
        @pkginfo.set['rhel']['7']['5.7']['client_package'] = %w(mysql-community-client mysql-community-devel)
        @pkginfo.set['rhel']['7']['5.7']['server_package'] = 'mysql-community-server'
        @pkginfo.set['smartos']['5.11']['5.5']['client_package'] = %w(mysql-client)
        @pkginfo.set['smartos']['5.11']['5.5']['server_package'] = 'mysql-server'
        @pkginfo.set['smartos']['5.11']['5.6']['client_package'] = %w(mysql-client)
        @pkginfo.set['smartos']['5.11']['5.6']['server_package'] = 'mysql-server'
        @pkginfo.set['suse']['11.3']['5.5']['client_package'] = %w(mysql-client)
        @pkginfo.set['suse']['11.3']['5.5']['server_package'] = 'mysql'

        @pkginfo
      end
    end

    def package_name_for(platform, platform_family, platform_version, version, type)
      keyname = keyname_for(platform, platform_family, platform_version)
      info = Pkginfo.pkginfo[platform_family.to_sym][keyname]
      type_label = type.to_s.gsub('_package', '').capitalize
      unless info[version]
        # Show availabe versions if the requested is not available on the current platform
        Chef::Log.error("Unsupported Version: You requested to install a Mysql #{type_label} version that is not supported by your platform")
        Chef::Log.error("Platform: #{platform_family} #{platform_version} - Request Mysql #{type_label} version: #{version}")
        Chef::Log.error("Availabe versions for your platform are: #{info.map { |k, _v| k }.join(' - ')}")
        fail "Unsupported Mysql #{type_label} Version"
      end
      info[version][type]
    end

    def keyname_for(platform, platform_family, platform_version)
      return platform_version if platform_family == 'debian' && platform == 'ubuntu'
      return platform_version if platform_family == 'fedora'
      return platform_version if platform_family == 'omnios'
      return platform_version if platform_family == 'rhel' && platform == 'amazon'
      return platform_version if platform_family == 'smartos'
      return platform_version if platform_family == 'suse'
      return platform_version.to_i.to_s if platform_family == 'debian'
      return platform_version.to_i.to_s if platform_family == 'rhel'
      return platform_version.to_s if platform_family == 'debian' && platform_version =~ /sid$/
      return platform_version.to_s if platform_family == 'freebsd'
    end

    def parsed_data_dir
      return new_resource.data_dir if new_resource.data_dir
      return "/opt/local/lib/#{mysql_name}" if node['os'] == 'solaris2'
      return "/var/lib/#{mysql_name}" if node['os'] == 'linux'
      return "/var/db/#{mysql_name}" if node['os'] == 'freebsd'
    end

    def client_package
      package_name_for(
        node['platform'],
        node['platform_family'],
        node['platform_version'],
        parsed_version,
        :client_package
        )
    end

    def server_package
      package_name_for(
        node['platform'],
        node['platform_family'],
        node['platform_version'],
        parsed_version,
        :server_package
        )
    end

    def server_package_name
      return new_resource.package_name if new_resource.package_name
      server_package
    end

    def parsed_version
      return new_resource.version if new_resource.version
      return '5.0' if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 5
      return '5.1' if node['platform_family'] == 'debian' && node['platform_version'] == '10.04'
      return '5.1' if node['platform_family'] == 'debian' && node['platform_version'].to_i == 6
      return '5.1' if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 6
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == '12.04'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == '13.04'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == '13.10'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == '14.04'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == '14.10'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'] == 'jessie/sid'
      return '5.5' if node['platform_family'] == 'debian' && node['platform_version'].to_i == 7
      return '5.5' if node['platform_family'] == 'fedora'
      return '5.5' if node['platform_family'] == 'freebsd'
      return '5.5' if node['platform_family'] == 'omnios'
      return '5.5' if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 2014
      return '5.5' if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 2015
      return '5.5' if node['platform_family'] == 'rhel' && node['platform_version'].to_i == 7
      return '5.5' if node['platform_family'] == 'smartos'
      return '5.5' if node['platform_family'] == 'suse'
    end
  end
end
